src/recordlayer.js

```diff
@@ -0,0 +1,329 @@
+
+// Encrypting at most 2^24 records will force us to stay
+// below data limits on AES-GCM encryption key use, and also
+// means we can accurately represent the sequence number as
+// a javascript double.
+const MAX_SEQUENCE_NUMBER = Math.pow(2, 24);
+const MAX_RECORD_SIZE = Math.pow(2, 14);
+const MAX_ENCRYPTED_RECORD_SIZE = MAX_RECORD_SIZE + 256;
+const RECORD_HEADER_SIZE = 5;
+
+// This is a helper class to manage the encryption/decryption state
+// for a particular key.
+
+export class CipherState {
+  constructor() {
+    this.key = null;
+    this.iv = null;
+    this.seqnum = 0;
+  }
+
+  async setKey(key, modes) {
+    // Derive key and iv per https://tools.ietf.org/html/rfc8446#section-7.3
+    this.key = await prepareKey(await hkdfExpandLabel(key, 'key', EMPTY, KEY_LENGTH), modes);
+    this.iv = await hkdfExpandLabel(key, 'iv', EMPTY, IV_LENGTH);
+    this.seqnum = 0;
+  }
+
+  nonce() {
+    // Ref https://tools.ietf.org/html/rfc8446#section-5.3:
+    // * left-pad the sequence number with zeros to IV_LENGTH
+    // * xor with the provided iv
+    const nonce = new Uint8Array(IV_LENGTH);
+    // Our sequence numbers are always less than 2^24, so fit in a Uint32.
+    new DataView(nonce.buffer).setUint32(IV_LENGTH - 4, this.seqnum);
+    for (let i = 0; i < IV_LENGTH; i++) {
+      nonce[i] ^= this.iv[i];
+    }
+    this.seqnum += 1;
+    if (this.seqnum > MAX_SEQUENCE_NUMBER) {
+      throw new TLSError(ALERT_DESCRIPTION.INTERNAL_ERROR);
+    }
+    return nonce;
+  }
+
+  async encrypt(plaintext, additionalData) {
+    return await encrypt(this.key, this.nonce(), plaintext, additionalData);
+  }
+
+  async decrypt(ciphertext, additionalData) {
+    return await decrypt(this.key, this.nonce(), ciphertext, additionalData);
+  }
+}
+
+// The main RecordLayer class.
+
+export class RecordLayer {
+  constructor(sendCallback) {
+    this.sendCallback = sendCallback;
+    this._sendCipherState = new CipherState();
+    this._recvCipherState = new CipherState();
+    this._sendError = null;
+    this._recvError = null;
+    this._pendingRecordType = 0;
+    this._pendingRecordBuf = null;
+  }
+
+  async setSendKey(key) {
+    await this.flush();
+    await this._sendCipherState.setKey(key, ['encrypt']);
+  }
+
+  async setRecvKey(key) {
+    await this._recvCipherState.setKey(key, ['decrypt']);
+  }
+
+  async setSendError(err) {
+    this._sendError = err;
+  }
+
+  async setRecvError(err) {
+    this._recvError = err;
+  }
+
+  async send(type, data) {
+    if (this._sendError !== null) {
+      throw this._sendError;
+    }
+    // Forbid sending data that doesn't fit into a single record.
+    // We do not support fragmentation over multiple records.
+    if (data.byteLength > MAX_RECORD_SIZE) {
+      throw new TLSError(ALERT_DESCRIPTION.INTERNAL_ERROR);
+    }
+    // Forbid sending plaintext application data.
+    if (this._sendCipherState.key === null && type === RECORD_TYPE.APPLICATION_DATA) {
```

Comment:
Note to self: there's a corresponding test to add here as well, that application-data records cannot be sent or received in the handshake phase.

Reply:
So, with the change to avoid buffering sent data during the handshake, the `TLSConnection.state` objects handle this state-management already.  For example all the handshake states will now throw if `recvApplicationData` or `sendApplicationData` is called on that state.  Is there still value in duplicating such checks in the record layer?